v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_PROPERTY` or `LOG_CONTEXT_STRING_PROPERTY` entries.
+
+Example:
+
+```c
+    LOG_CONTEXT_LOCAL_DEFINE(log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow writing code like below while not performing the memory copies until the logging is actually executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOG_CONTEXT_LOCAL_DEFINE(local_context, NULL, LOG_CONTEXT_STRING_PROPERTY(correlation_id, "%s", correlation_id));
+
+    if (malloc(42) == NULL)
+    {
+        logger_log(LOG_LEVEL_ERROR, local_context, "malloc failed");
+    }
+}
+```
+
+### Dynamically allocated context creation
+
+The library shall support creating a dynamically allocated context.
+
+```c
+#define LOG_CONTEXT_CREATE(log_context_handle, parent_context, ...) \
+
+#define LOG_CONTEXT_CREATE(log_context_handle) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_STRING_PROPERTY`.
+
+Example:
+
+```c
+    LOG_CONTEXT_CREATE(dynamically_allocated_log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+    LOG_CONTEXT_DESTROY(dynamically_allocated_log_context);
+```
+
+### Context chaining
+
+It shall be supported to chain contexts (define a context or create a context dynamically while specifying a parent context to inherit the information from).
+
+### LOG_CONTEXT_PROPERTY
+
+`LOG_CONTEXT_PROPERTY` is a macro that allows defining non-string properties by specifying a property type, name and a value.
+
+```c
+#define LOG_CONTEXT_PROPERTY(property_type, property_name, property_value) \
+    ...
+```
+
+The following types shall be supported out of the box:
+
+- ansi_charptr
+- wchar_t_ptr
+- int8_t
+- uint8_t
+- int16_t
+- uint16_t
+- int32_t
+- uint32_t
+- int64_t
+- uint64_t
+
+Note: more types will be added as needed.
+
+### LOG_CONTEXT_STRING_PROPERTY
+
+`LOG_CONTEXT_STRING_PROPERTY` is a macro that allows defining context properties by formatting the value as a string with `printf`-like formatting.
+
+```c
+#define LOG_CONTEXT_STRING_PROPERTY(property_name, property_format_specifier, property_value) \
+    ...
+```
+
+### LOG_CONTEXT_NAME
+
+```
+#define LOG_CONTEXT_NAME(context_name) \
+```
+
+`LOG_CONTEXT_NAME` shall allow specifying the context name to be used either when printing logs to console or when ETW events are published.
+
+By default if not specified, the context name shall be `unnamed`.
```

Comment:
>be `unnamed`. [](http://example.com/codeflow?start=52&length=13)

"not be part of the log line"?

Reply:
I changed this to be an empty string and it should not show anymore in the console logging. How it gets displayed by various ETW viewers is a different thing.